bumble/profiles/hap.py

```diff
@@ -0,0 +1,599 @@
+            ),
+        )
+        self.active_preset_index_characteristic = gatt.Characteristic(
+            uuid=gatt.GATT_ACTIVE_PRESET_INDEX_CHARACTERISTIC,
+            properties=(
+                gatt.Characteristic.Properties.READ
+                | gatt.Characteristic.Properties.NOTIFY
+            ),
+            permissions=gatt.Characteristic.Permissions.READ_REQUIRES_ENCRYPTION,
+            value=bytes([self.active_preset_index]),
+        )
+
+        super().__init__(
+            [
+                self.hearing_aid_features_characteristic,
+                self.hearing_aid_preset_control_point,
+                self.active_preset_index_characteristic,
+            ]
+        )
+
+    # TODO this need to be triggered when device is unbonded
+    def on_forget(self, addr: Address):
+        self.preset_changed_operations_history_per_device.pop(addr)
+
+    def _on_write_hearing_aid_preset_control_point(
+        self, connection: Optional[Connection], value: bytes
+    ):
+        assert connection
+
+        opcode = HearingAidPresetControlPointOpcode(value[0])
+        handler = getattr(self, '_on_' + opcode.name.lower())
+        connection.abort_on('disconnection', handler(connection, value))
+
+    async def _on_read_presets_request(
+        self, connection: Optional[Connection], value: bytes
+    ):
+        assert connection
+
+        if self.read_presets_request_in_progress:
+            raise att.ATT_Error(CommonErrorCode.PROCEDURE_ALREADY_IN_PROGRESS)
+        self.read_presets_request_in_progress = True
+
+        start_index = value[1]
+        if start_index == 0x00:
+            raise att.ATT_Error(CommonErrorCode.OUT_OF_RANGE)
+
+        num_presets = value[2]
+        if num_presets == 0x00:
+            raise att.ATT_Error(CommonErrorCode.OUT_OF_RANGE)
+
+        # Sending `num_presets` presets ordered by increasing index field, starting from start_index
+        presets = [
+            self.preset_records[key]
+            for key in sorted(self.preset_records.keys())
+            if self.preset_records[key].index >= start_index
+        ]
+        del presets[num_presets:]
+        if len(presets) == 0:
+            raise att.ATT_Error(CommonErrorCode.OUT_OF_RANGE)
+
+        AsyncRunner.spawn(self._read_preset_response(connection, presets))
+
+    async def _read_preset_response(
+        self, connection: Connection, presets: List[PresetRecord]
+    ):
+        try:
+            while len(presets) > 0:
+                await connection.device.indicate_subscriber(
+                    connection,
+                    self.hearing_aid_preset_control_point,
+                    value=bytes(
+                        [
+                            HearingAidPresetControlPointOpcode.READ_PRESET_RESPONSE,
+                            len(presets) == 1,
+                        ]
+                    )
+                    + bytes(presets[0]),
+                )
+                presets.pop(0)
```

Comment:
nit: In Python I will prefer for-loop over while-loop

```suggestion
            for i, preset in enumerate(presets):
                await connection.device.indicate_subscriber(
                    connection,
                    self.hearing_aid_preset_control_point,
                    value=bytes(
                        [
                            HearingAidPresetControlPointOpcode.READ_PRESET_RESPONSE,
                            i == len(presets) - 1,
                        ]
                    )
                    + bytes(preset),
                )
              presets.clear()
```

Reply:
Calling pop for each indicate allow to remove the notification that has been sent only.

If we do the for loop and call clear, we are not handling the case in which the remove device doesn't received an indicate.
In such scenario, the service should try to resend the indication when it reconnect later

Reply:
I misunderstood your comment, thinking it was attached to _preset_changed_operation.

I don't think using a for loop is good for the _preset_changed_operation, but it is definitively possible for the read preset one